### ISSUE TITLE
Threading model discussion

### ISSUE THREAD

**Issue by tomaka** [tomaka/glium#459]
## How OpenGL works

To do something with OpenGL, there are three steps:

 - Create the context
 - Set the context as being the active context in the current thread
 - Call the `glSomething` functions

Where a function in a regular C library would look like `do_something(const Context* context, int param);`, OpenGL instead looks like `do_something(int param);` and operates on the context that is active in the current thread.

Additional restrictions: you can only have one active context per thread, and the same context can't be active in two different threads at the same time (though this sometimes works depending on the backend).

Also on windows it is **very** slow to change the current context. So we should avoid this too.

## How Glium currently handles this

Glium creates an internal background thread that is inaccessible from outside the library. The context is created by glium and set as current in this background thread. This ensures that the context isn't active anywhere else, and that the current context of the thread is never changed.

Then all function calls add a command to a queue that is executed by this background thread.

This approach has the following drawbacks:

- It is slower than raw OpenGL calls.
- Embedded platforms often can't afford an additional thread.
- Having hidden costs is not what you expect from a system language.

## What's the alternative?

Glium aims to be 100% safe, which is an additional constraint in the design.

There are two problems when you want to remove this background thread:

- We must ensure that we are on the right thread when calling OpenGL functions
- We must ensure that the current context didn't change

The first problem is easily solved by removing `Send` and `Sync` from *all* structs that represent an OpenGL object (the display, textures, buffers, etc.).

This has the disadvantage that glium users must do everything from within the same thread, possibly reimplementing something similar to the commands queue that currently exists in glium.

The second problem is a bit more annoying:

 - The user can call `MakeCurrent` manually, but the function is marked as unsafe so we can ignore this situation.
 - The user can create a second context in the same thread.
 - The user can use another OpenGL wrapper which also exposes a safe interface, and create a context in the same thread.

The only solution that I see to make this safe is calling `*GetCurrentContext()` before each OpenGL function call. I don't know how this affects performances.

## TL;DR

There are two alternatives: 

- Having a background thread (the current one)
- Not implementing Send+Sync and calling `GetCurrentContext()` all the time

**Comment by tomaka** [comment]
I'd like to get some opinions about: would this be a problem for you to do all the rendering from the same thread, and having all glium objects not implement Send/Sync?

I might also wait until we get more information about glNext. If Glium can be made compatible with glNext, glNext has threading issue in its API, and has a pretty good coverage, then I may as well keep the background thread for legacy OpenGL and use direct calls for glNext.

**Comment by tomaka** [comment]
Now that we know for sure that ~~glNext~~ Vulkan is a totally different API that can't be made compatible with glium, I'm in favor or being closer to OpenGL and forcing everything to be single-threaded.

What this would change: all objects built from glium would need to be on the same thread. This means that for example if you want to load a texture from another thread you'll have to transfer its data to the main thread first. This is already what happens under the hood in glium.
